### ISSUE TITLE
when will fully code and weights release?

### ISSUE THREAD

**Issue by lucasjinreal** [postech-ami/FastMETRO#3]
v

**Comment by FastMETRO** [comment]
Hello,

We are planning to release our complete code and pre-trained models on October or November.

Thanks for your interest in our work!!

----

Our code is under refactoring. In the meantime, please refer to [[Code Snippet (Overall Architecture)]](https://github.com/postech-ami/FastMETRO/blob/main/src/fastmetro.py) and [[Code Snippet (Transformer Encoder-Decoder)]](https://github.com/postech-ami/FastMETRO/blob/main/src/transformer.py) of our FastMETRO.

The training procedure for FastMETRO is based on [METRO](https://github.com/microsoft/MeshTransformer). It would be helpful for you to check [[INSTALL]](https://github.com/microsoft/MeshTransformer/blob/main/docs/INSTALL.md), [[DOWNLOAD]](https://github.com/microsoft/MeshTransformer/blob/main/docs/DOWNLOAD.md) and [[EXP]](https://github.com/microsoft/MeshTransformer/blob/main/docs/EXP.md) provided by [METRO](https://github.com/microsoft/MeshTransformer).

**Comment by FastMETRO** [comment]
Please reopen this issue if you need more help regarding this.

**Comment by lucasjinreal** [comment]
@FastMETRO thanks, hoping for it. Meanwhile, may I ask if FastMETRO can runing realtime along with a detector (yolov5) in realworld applications?

**Comment by FastMETRO** [comment]
@jinfagang  

Regarding the inference time, FastMETRO-S takes about 9.6ms and ResNet50 takes about 7.5ms in our experiment. Since the overall inference time is only about 17.1ms (58.5FPS), we can deploy our model for real-world applications. Note that we measured the inference time using a single V100 GPU with a batch size of 1 on [Human3.6M](http://vision.imar.ro/human3.6m/pami-h36m.pdf). 

In addition to the overall inference time of our model, we also need to consider the inference time of a detector (e.g., yolov5), because FastMETRO assumes the input image where the target person is located at the center.

---

In terms of transformer architectures, FastMETRO-S is about **_10× smaller and 2.5× faster_** than [METRO](https://arxiv.org/abs/2012.09760) and [Mesh Graphormer](https://arxiv.org/abs/2104.00272).

<img width="724" alt="table2" src="https://user-images.githubusercontent.com/109520798/186669712-c4c84bc2-4980-4cf0-8d8f-5c2cb19c3758.png">

**Comment by lucasjinreal** [comment]
@FastMETRO that's promising, thanks for this nice work! I have one last question:

Have FastMETRO compared with PARE? how's the transition error compare with it? (since in real world, we might need the 3d world poses as well as pred_cams as transitions, etc, in a mocap system)

**Comment by FastMETRO** [comment]
@jinfagang, thanks for the kind words!!

As shown in Table 3, we compared FastMETRO with PARE:

<img width="726" alt="table3" src="https://user-images.githubusercontent.com/109520798/186727459-f3d11537-28c0-4d8e-8330-f4f29c74fdc4.png">

###### * R50: ResNet-50   |   H32: HRNet-W32   |   H64: HRNet-W64

When both models employ the same CNN backbone (R50), FastMETRO-S achieves higher accuracy. In other words, FastMETRO-S-R50 outperforms PARE-R50. 

Since it is difficult to compute the absolute 3D locations in the 3D world from a single RGB image, our model only estimates weak-perspective camera parameters. Nevertheless, we could approximate what you mentioned. Please refer to these great discussions from Prof. Kanazawa ([akanazawa/hmr#81](https://github.com/akanazawa/hmr/issues/81)) and from Dr. Lin ([microsoft/MeshTransformer#11](https://github.com/microsoft/MeshTransformer/issues/11)).

**Comment by lucasjinreal** [comment]
@FastMETRO thanks! From my personal expreiences, PARE predicts pred_cams params as transition very jitting and shaking ( weakly-perspective camera also should be ok since am just using it apply on mocap model), but in temporal sequence, one frame between next another are shaking would cause some un-netural problem.

Does FastMETRO also have pred_cams as PARE does? does it have a better consistency compare with PARE?

**Comment by FastMETRO** [comment]
@jinfagang 

FastMETRO also estimates weak-perspective camera parameters (`cam_parameter`  in [fastmetro.py](https://github.com/postech-ami/FastMETRO/blob/main/src/fastmetro.py)); a scaling factor `s` and 2D translation vector `(tx, ty)`.

Since our method reconstructs 3D human mesh from a single input image, temporal consistency is not considered and not measured in our evaluations; FastMETRO belongs to image-based methods (e.g., [GraphCMR](https://arxiv.org/abs/1905.03244)). Thus, it is hard to fairly compare FastMETRO with PARE in terms of the temporal consistency. 

To achieve better temporal consistency, we might need to exploit consecutive frames; this belongs to video-based methods (e.g., [VIBE](https://arxiv.org/abs/1912.05656)). Exploiting consecutive frames with our model architecture is an interesting future work.

**Comment by lucasjinreal** [comment]
@FastMETRO Indeed. But on the other hand, since we can make the Projection error as little as possible, the mesh should match original location as fit as possible. which means, the projection is correct enough, then [scale, tx, ty] should also be more accurate. 
at least tx ty is more accurate, do u know a way to recover 3d position from [scale, tx, ty] more accurate if we might not have a accurate scale?

**Comment by FastMETRO** [comment]
@jinfagang 

### [3D Human Mesh Output]
Our model directly estimates the 3D coordinates of human mesh vertices. Thus, even though the weak-perspective camera parameters are not accurate, we still could obtain accurate 3D human mesh output.


### [2D Human Mesh Output]
We project the 3D human mesh output onto the 2D input image by using the weak-perspective camera parameters via orthographic projection. So, if these parameters are not accurate, then the projected 2D human mesh would be inaccurate.

---

As described in [3D Human Mesh Output], the accuracy of weak-perspective camera parameters (i.e., the accuracy of a scaling factor `s` and the accuracy of a 2D translation vector `(tx, ty)`) does not affect the accuracy of 3D human mesh output.

**Comment by lucasjinreal** [comment]
@fastMETRO I just noticed that fastMetro just like mesh former, which outputs 3d verts directly, rather than betas and shape params of SMPL, 
but there is a very important thing, in meshformer it 's output has some spikes on mesh, does fastMETRO have this issue?

**Comment by FastMETRO** [comment]
@jinfagang 

Generally speaking, there are two approaches in the literature: parametric approach (which estimates model parameters of a human body model) and non-parametric approach (which estimates 3D coordinates of human mesh vertices). Our method belongs to the non-parametric approach.

As in other methods in this approach (e.g., [GraphCMR](https://arxiv.org/abs/1905.03244) and [METRO](https://arxiv.org/abs/2012.09760)), the 3D human mesh output estimated by FastMETRO could have some spikes on the mesh. This issue could be alleviated by training our model with the mesh normal vector loss and the mesh edge length loss. We are planning to apply these losses for the training of our FastMETRO.

We could estimate model parameters of a human body model by using the estimated 3D human mesh vertices. Following [GraphCMR](https://arxiv.org/abs/1905.03244), we employed fully connected layers as a SMPL parameter regressor and visualized the results in Figure C2.

<img width="719" alt="smpl_regressor" src="https://user-images.githubusercontent.com/109520798/187081506-e86856cf-f732-4a42-bcf2-f26a462e9591.png">

**Comment by lucasjinreal** [comment]
@FastMETRO  Glad fasterMETRO actually noticed this issue and fixed it. 

It looks like fastMETR able to predict paramatric of SMPL as well as directly mesh prediction? (correct me if am wrong)

Then, it makes me have a further question: Directly predict mesh have a huge output tensor which is hard to regress and has spikes issue as described above, so why do we want this way though? Does  it has some strengths that parametric model doesn't have?

**Comment by FastMETRO** [comment]
@jinfagang 

FastMETRO directly estimates 3D coordinates of human mesh vertices. Optionally, we can also predict pose and shape coefficients of SMPL; a SMPL parameter regressor takes the estimated 3D vertex coordinates as input, then estimates the pose and shape coefficients of SMPL.

One big advantage of the non-parametric approach is that the regression targets are 3D coordinates. It is reported that the regression of 3D rotations (e.g., pose coefficients of SMPL) is very difficult, but the 3D rotations are the regression targets in the parametric approach. Due to this challenge, the parametric approach usually exploits test-time optimizations or iterative fitting procedures.

**Comment by lucasjinreal** [comment]
@FastMETRO do u mean fastMETR predict 3d mesh more intuitive and more accurate in 3d positions? In mesh prediction method how to get the joints 3d positions and rotations from predict meshs?

**Comment by Youwang-Kim** [comment]
@jinfagang 
Hi, I am the 2nd author of FastMETRO. Thanks for your interest in our work and overall human mesh estimation. 

In usual mesh 'parameter' prediction methods, we usually regress SMPL mesh parameters (via CNNs or else), and we utilize the pre-defined linear blend skinning function and 3D joint regression matrix of SMPL. This approach is quite basic in human mesh estimation work using SMPL models. 

Please refer to this [link](https://github.com/vchoutas/smplx/blob/5fa20519735cceda19afed0beeabd53caef711cd/smplx/body_models.py#L306) for more details.

**Comment by lucasjinreal** [comment]
@Youwang-Kim I think you didn't get my point. From my discussion with @FastMETRO (not sure if you are same person or not). My point is FastMETRO directly predict mesh rather than predict co-efficient as PARE or VIBE does. And FastMETRO indicates directly predict mesh can have a better precision on 3D positions (correct me if am wrong).

From my point of view, if 3D position is better than PARE or VIBE, this is indeed a advantage of `non-parmametric` method.
So further, how will this mesh get 3d joints?

**Comment by Youwang-Kim** [comment]
@jinfagang 
First, as you mentioned, our FastMETRO directly predicts 3D human mesh vertices, rather than SMPL parameters. Please note that the vertices we predict, have the same topology with the SMPL models, thus, we can still utilize SMPL's 3D joint regression matrix (vertices -> 3D joints) even though FastMETRO is non-parametric method. Please refer to previous non-parametric methods such as METRO, Mesh Graphormer and GraphCMR. Such non-parametric methods utilize SMPL's mesh topology, so they are all capable of regressing 3D joints from the predicted vertices. 

Next, one cannot generalize the non-parametric methods are always better than parametric methods. As you mentioned, non-parametric methods show better reconstruction accuracy, recently. You can check I2L-MeshNet (ECCV2020), where they claim as below.

> "However, directly regressing the parameters from the input image is a highly non-linear mapping because it breaks the spatial relationship between pixels in the input image." 

In our paper, we show that FastMETRO's cross attention of 2D image and 3D mesh vertex/joint tokens are important for mitigating such issues, and help achieving improved accuracy, faster convergence and lightweight models. Note that we are not generalizing non-parametric methods are always better than parametric methods. 

Hope this comment helps.

**Comment by lucasjinreal** [comment]
@Youwang-Kim thanks it's helpful.

Looking forward to your guys release training code, seems benefits from transformer directly predict mesh is a promising way. Old methods like PARE and VIBE regression meshs good enough, but transition always shaking. This might caused by:  it breaks the spatial relationship between pixels in the input image
